### PR TITLE
fix(api_core): improve rest path validation

### PR DESCRIPTION
**Pull request by daniel-sanche**
Context: b/532254612

### CODE REVIEW

File: packages/google-api-core/google/api_core/path_template.py [hunk 1]
```diff
@@ -62,6 +63,179 @@
...
+        for i, wildcard_type in enumerate(wildcard_types):
+            captured_val = m.group(i + 1)
+            if wildcard_type == "*":
+                if captured_val in (".", ".."):
+                    name_str = property_name if property_name else "positional variable"
+                    raise ValueError("Invalid value {} for {}.".format(val, name_str))
+            else:
+                if not _validate_multi_segment_value(captured_val):
+                    name_str = property_name if property_name else "positional variable"
+                    raise ValueError("Invalid value {} for {}.".format(val, name_str))
```
[the remark this answers is not in the document]
Comment (daniel-sanche) [reply packages/google-api-core/google/api_core/path_template.py]: This should not be strictly needed, since the values are later compared against the templates in the validate() step. But I added this change anyway

File: packages/google-api-core/google/api_core/path_template.py [hunk 2]
```diff
@@ -62,6 +63,189 @@
...
+    Args:
+        val (str): The raw string value to validate.
+        template_str (str): The template string of the variable (e.g. 'projects/*/locations/*').
+        property_name (str | None): The name of the property being validated, used
+            to construct descriptive error messages.
+
+    Raises:
+        ValueError: If a wildcard within a structurally valid value violates path traversal rules.
+    """
+    err = ValueError(
```
Comment (chalmerlowe) [inline packages/google-api-core/google/api_core/path_template.py]: #NIT 
I do not know how often `_extract_and_validate_wildcards` gets called. IF it is regular occurrence we will be executing the `f-string` and `ValueError` creation every single time we call the function, even when we don't have an error condition.

If we call the function often AND errors are expected to be rare then a factory function (`def create_err_msg()`) to generate an err message only when we raise (a la: `raise create_err_msg()`) could be useful.

File: packages/google-api-core/google/api_core/path_template.py [hunk 3]
```diff
@@ -62,6 +63,189 @@
...
+        f"Invalid value {val} for {property_name or 'positional variable'}."
+    )
+
+    # Single-segment templates (None or "*") cannot match exactly "." or ".."
+    # and cannot have multi-segment paths resolving to 0 segments.
+    if template_str is None or template_str == "*":
+        if val in (".", "..") or (val and not _validate_multi_segment_value(val)):
```
Comment (chalmerlowe) [inline packages/google-api-core/google/api_core/path_template.py]: #PREFERENCE/QUESTION:
Could we benefit from changing the order of operations?

If the REST API is a standard web service, 99.9% of incoming requests will be normal path segments (like "api", "v1", "users", "12345"). They will rarely be "." or "..".

This means our first check (the tuple check) will "miss" 99.9% of the time, forcing Python to execute the `_validate_multi_segment_value` anyway.

We can bypass the tuple lookup entirely by rewriting the logic. Since we know that `_validate_multi_segment_value(val)` already natively returns `False` for dots ("." OR ".."), we can drop the tuple check entirely to save a few cycles on every normal path

```python
# Optimized for the 99.9% use-case (normal path segments)

if val and not _validate_multi_segment_value(val):
```

Why this is faster for REST routing:

**Old Way (Normal Path)**: Checks tuple (Misses a lot), Checks if val (True), then Runs the function

**New Way (Normal Path)**: Checks val (True) then Runs the Function.

File: packages/google-api-core/google/api_core/path_template.py [hunk 4]
```diff
@@ -62,6 +63,189 @@
...
+            raise err
+        return
+
+    # Multi-segment templates ("**") must represent at least one valid, non-escaped segment.
+    if template_str == "**":
```
Comment (chalmerlowe) [inline packages/google-api-core/google/api_core/path_template.py]: #QUESTION

There is a subtle inconsistency in how we handle empty strings between the two `if` statements and the `else` statement. Is this intentional?

#### Snippet 1: Handling Single Segments (`*`)

This block checks if the template is for a single segment (like `*`).

```python
    # Single-segment templates (None or "*") cannot match exactly "." or ".."
    # and cannot have multi-segment paths resolving to 0 segments.
    if template_str is None or template_str == "*":
        if val in (".", "..") or (val and not _validate_multi_segment_value(val)):
            raise err
        return
```

**What happens if `val` is an empty string (`""`)?**
The check `val in (".", "..")` is **False**.
Then it checks `(val and not _validate_multi_segment_value(val))`. Since `val` is `""` (which is "falsy" in Python), the `val and` part evaluates to **False**. 
The whole condition immediately evaluates to False, so it **bypasses** the error and returns successfully. **An empty string is allowed here.**

#### Snippet 2: Handling Bare Multi-Segments (`**`)

This block checks if the template is a bare double star (like `**`).

```python
    # Multi-segment templates ("**") must represent at least one valid, non-escaped segment.
    if template_str == "**":
        if not _validate_multi_segment_value(val):
            raise err
        return
```

**What happens if `val` is an empty string (`""`)?**
There is no `val and` guard here. It calls `_validate_multi_segment_value("")` directly. 
If we trace `_validate_multi_segment_value("")`, it splits the empty string into segments, ends up with 0 leftover segments, and returns `False`.
So `not _validate_multi_segment_value("")` becomes `not False` which is **True**.
**So this snippet raises an error. An empty string is rejected here.**

#### Snippet 3: Falling back for Complex Templates

This block is at the very end of the function, handling cases where the value didn't match the expected complex pattern (e.g. something like `projects/{project}/locations/{location}`).

```python
    else:
        # For values that don't match the pattern, ensure the value doesn't
        # resolve to 0 segments (e.g. "projects/..").
        if val and not _validate_multi_segment_value(val):
            raise err
```

**What happens if `val` is an empty string (`""`)?**
Just like Snippet 1, we have a `val and` guard here.
Since `val` is `""` (falsy), the condition `val and ...` is **False**.
**It bypasses the error. An empty string is allowed here.**

### Why this is inconsistent:

If a user passes an empty string `""` as the value:
*   If the template is `*` (Snippet 1), it **passes**.
*   If the template is `**` (Snippet 2), it **fails**.
*   If the template is a complex pattern (Snippet 3), it **passes**.

Is this difference in behavior intentional?

File: packages/google-api-core/google/api_core/path_template.py [hunk 5]
```diff
@@ -62,6 +63,189 @@
...
+            - The compiled regex pattern string with capture groups.
+            - A list of wildcard type strings ('*' or '**') in matching order.
+    """
+    wildcard_types = []
+
+    def replacer(match):
+        positional = match.group("positional")
+        template = match.group("template")
+
+        if positional == "*":
```
Comment (chalmerlowe) [inline packages/google-api-core/google/api_core/path_template.py]: #PREFERENCE: Above we define constants that can be applied to the return values for each of these conditions. I recommend we use them:

```
# Segment expressions used for validating paths against a template.
_SINGLE_SEGMENT_PATTERN = r"([^/]+)"
_MULTI_SEGMENT_PATTERN = r"(.+)"
```

File: packages/google-api-core/tests/unit/test_path_template.py [hunk 6]
```diff
@@ -650,3 +674,181 @@ def helper_test_transcode(http_options_list, expected_result_list):
...
+@pytest.mark.parametrize(
+    "val, expected_valid",
+    [
+        ("a/b/c", True),
+        ("a/../b", True),
+        ("a/b/c/d/e/../../../..", True),
+        ("a/b/../..", False),
+        ("a/b/../../..", False),
+        (".", False),
+        ("..", False),
```
Comment (chalmerlowe) [inline packages/google-api-core/tests/unit/test_path_template.py]: #PREFERENCE:

I recommend some additional checks to cover some more edge cases:

```suggestion
        ("..", False),
        ("", False),
        ("/", False),
        ("//", False),
        ("a/../../b", False),
        ("../..", False),
```

File: packages/google-api-core/tests/unit/test_path_template.py [hunk 7]
```diff
@@ -650,3 +674,181 @@ def helper_test_transcode(http_options_list, expected_result_list):
...
+    "template_str, expected_pattern, expected_wildcards",
+    [
+        (
+            "projects/{project}/locations/{location}",
+            "projects/([^/]+)/locations/([^/]+)",
+            ("*", "*"),
+        ),
+        ("projects/{project=**}", "projects/(.+)", ("**",)),
+        ("projects/{project=locations/*}", "projects/locations/([^/]+)", ("*",)),
+        ("projects/*/locations/**", "projects/([^/]+)/locations/(.+)", ("*", "**")),
```
Comment (chalmerlowe) [inline packages/google-api-core/tests/unit/test_path_template.py]: QUESTION: What if the template is just "projects/my-project" with no wildcards at all? It should return the literal escaped and an empty tuple of wildcard types. Do you think we should add the sanity check to ensure the regex parser doesn't choke on plain strings?

("projects/my-project", "projects/my-project", ()),

**Review by chalmerlowe — CHANGES_REQUESTED** [review]
Some comments and suggestions.

File: packages/google-api-core/google/api_core/path_template.py (hunk 5 above)
Comment (daniel-sanche) [reply packages/google-api-core/google/api_core/path_template.py]: Good catch, fixed

File: packages/google-api-core/google/api_core/path_template.py (hunk 3 above)
Comment (daniel-sanche) [reply packages/google-api-core/google/api_core/path_template.py]: Good catch, I removed the redundant checks

File: packages/google-api-core/tests/unit/test_path_template.py (hunk 7 above)
Comment (daniel-sanche) [reply packages/google-api-core/tests/unit/test_path_template.py]: added

File: packages/google-api-core/google/api_core/path_template.py (hunk 4 above)
Comment (daniel-sanche) [reply packages/google-api-core/google/api_core/path_template.py]: > This block is at the very end of the function, handling cases where the value didn't match the expected complex pattern (e.g. something like projects/{project}/locations/{location}).

The high-level answer to your question is that these functions are checking for security issues, not doing generic validation against the templates.The actual validation against the regex happens later in the stack, in the [`validate()` step](https://github.com/googleapis/google-cloud-python/blob/81482ba5e268310273958756b7d38d0287da7d29/packages/google-api-core/google/api_core/path_template.py#L225).

The reason we have two phases is that an API can provide multiple templates to check against (e.g. `parent=(projects/** | cluster/**)`. So if we encounter an input that doesn't line up with a template in the `validate()` step, we just go through the list and find another template to match. We don't want to raise an exception unless we have to.

But the code is in this PR is checking for security issues. If we find any input that "breaks free" from any template, we want to reject it immediately by raising an exception, and don't keep looking for other inputs that could accept it

The * and ** segment have different security implications, that are detailed in the doc I sent you. Basically, if we encounter an empty string for *, we can keep going and let `validate()` decide whether to accept the input or continue to the next template. But if we encounter a ** segment that resolves to an empty path, we need to treat that as a potential security issue

Comment (daniel-sanche) [reply packages/google-api-core/google/api_core/path_template.py]: I tried to capture some of this [in the docstrings](https://github.com/googleapis/google-cloud-python/blob/0d6d1ec5966b6f316e1a346ab0c6c8be38bac909/packages/google-api-core/google/api_core/path_template.py#L185), but let me know if you have suggestions for how to word this better

File: packages/google-api-core/google/api_core/path_template.py (hunk 2 above)
Comment (daniel-sanche) [reply packages/google-api-core/google/api_core/path_template.py]: I fixed this to only build an exception after an error is encountered

**Comment by daniel-sanche** [comment]
Converting this to draft, because it sounds like there are still some cross-language discussions happening

File: packages/google-api-core/google/api_core/path_template.py (hunk 4 above)
Comment (daniel-sanche) [reply packages/google-api-core/google/api_core/path_template.py]: Resolving this, since the design changed substantially

**Review by hebaalazzeh — APPROVED** [review]
LGTM

**Review by chalmerlowe — APPROVED** [review]
LGTM
